### DISCUSSION TITLE
Issues with multiple concurrent users

### DISCUSSION THREAD

**Discussion by JulieMatic** [artilleryio/artillery#1711] [Artillery]
I am currently working on some tests that are using artillery-engine-playwright, and I'm running into a couple of different issues depending on the combination of versions I use. 

### Only the "active" Chrome window is interacted with
```
phases:
            - name: 10user
              duration: 300
              arrivalRate: 1
              arrivalCount: 1000
              maxVusers: 10
```
Using the above phase, artillery version 1.7.9, and artillery-engine-version 0.2.0, when I run `artillery run --environment 10_concurrent_users ./dist/tests/Performance/artillery/artillery-10-users-prod-intro.yaml`, over the course of a few seconds, I see 10 Chrome windows pop up, but only the "active" window gets interacted with. For example, the first thing my test does is enter an email address, password, and click "Sign in." 9 of the windows do nothing, and the "active" window will input multiple email addresses into the email field. For all 10 windows, the tests get stuck at the login page because either no credentials are entered, or invalid credentials are provided. 

_Note_: So that I can see what's happening more easily, I'm running with `headless: false`, but one of my coworkers seems to be running into a similar issue where the tests get stuck at the login page and he is setting `headless: true`

### Phase configuration not behaving the same as before
```
phases:
            - name: 10user
              duration: 300
              arrivalRate: 1
              arrivalCount: 1000
              maxVusers: 10
```
Using the above phase (same as the previous phase), artillery version 2.0.0-27, and artillery-engine-version 0.2.1, when I run the same `artillery run` command, I only see 1 user running at a time, despite `maxVusers` being set to 10. 

If I instead configure my phase as follows:
```
phases:
            - name: 10user
              duration: 300
              arrivalRate: 3
              maxVusers: 10
```
When I run the same `artillery run` command, I initially see 2 Chrome windows pop up, the tests for those 2 windows complete, and then 3 more windows pop up. Then, throughout the duration of the artillery test, the maximum number of concurrent users I see at any given time is 3. With this phase, I'm seeing the issue where the test is only interacting with the active issue go away, and the playwright scripts can finish as expected. 

Using a similar phase configuration, but changing `arrivalRate` to 10, when I run the same `artillery run` command, I get something like 4 Chrome windows popping up, the tests for those windows completing, and then a couple more Chrome windows being added as tests finish, until I reach the maximum of 10. 

This is leading to me to believe that something changed with the phase configuration options. My understanding of the phase configuration options (based on [this documentation](https://www.artillery.io/docs/guides/guides/test-script-reference)) is that:

- `arrivalRate` is used to specify how many virtual users to add each second, so if I have `duration` set to 300 and `arrivalRate` set to 3, I should expect to see 3 virtual users added every second for 5 minutes
- `arrivalCount` is used to specify how many virtual users to add over the duration, so if I have `duration` set to 300 and `arrivalCount` set to 1000, I'd expect to see 1 virtual user about every 3 seconds 
- I haven't been able to figure out if `arrivalRate` and `arrivalCount` can/should be used together

But instead, it seems like `arrivalRate` is behaving like `maxVusers` and that there is a built-in ramp-up behavior. 

### Questions

1. Is there a known issue with artillery-engine-playwright where running artillery with multiple concurrent virtual users will result in only the active window being interacted with?
2. Was there a change in the phase configuration options? Without changing my YAML file between switching versions, I'm seeing a difference in behavior between artillery version 1.7.9 and artillery version 2.0.0-27, but I'm not sure which version is behaving correctly. 
3. Using the phases I provided in this post as examples, what is the expected behavior?
4. What is the ideal combination of versions for artillery and artillery-engine-playwright?

**Comment by juanig1** [comment]
Hey, thank you for the thorough report and sorry for the delay!

Been looking into this yesterday. I'll go in order:

> I haven't been able to figure out if `arrivalRate` and `arrivalCount` can/should be used together
You should use one or the other, if both are present `arrivalCount` takes precedence. If possible i would favor `arrivaRate` as it uses resources better. 

Regarding your questions:
> Is there a known issue with artillery-engine-playwright where running artillery with multiple concurrent virtual users will result in only the active window being interacted with?

There isn't, I can run some tests regarding this issue and let you know.

>Was there a change in the phase configuration options? Without changing my YAML file between switching versions, I'm seeing a difference in behavior between artillery version 1.7.9 and artillery version 2.0.0-27, but I'm not sure which version is behaving correctly.

Diving into what you shared I found a bug that's causing `maxVusers` to generate less vusers than expected. Depending on the amount of threads you workstation has it could be around 1 or 2, so that may explain you seeing only one user.
Working on a fix right now!

> Using the phases I provided in this post as examples, what is the expected behavior?
Taking this one as an example:

```
phases:
            - name: 10user
              duration: 300
              arrivalRate: 3
              maxVusers: 10
```
300s phase with 3 new vusers per second. When trying to create a new user, if there are already 10 concurrent vusers, it is skipped (you should see a log line regarding skipped users). This can happen if vusers take more than 1s to finish (which should be the case i think when using playwright).
> What is the ideal combination of versions for artillery and artillery-engine-playwright?

Latest is encouraged, so: `2.0.0-27` and `0.2.1` respectively. 

Let me work on the `maxVusers` fix and I'll update here when I have more info.
Feel free to ask anything that is not clear in my response!

**Reply by JulieMatic** [reply]
Thank you so much for the thorough response! It was very helpful. 

> You should use one or the other, if both are present arrivalCount takes precedence. If possible i would favor arrivaRate as it uses resources better.

I will go ahead and update our YAML files to specify `arrivalRate` for our phases, as I think it is more inline with what we're trying to test.

> Diving into what you shared I found a bug that's causing maxVusers to generate less vusers than expected. Depending on the amount of threads you workstation has it could be around 1 or 2, so that may explain you seeing only one user.

Yes, I think that does explain why I'm only seeing 1 user. When my coworker ran with the same phase on a Kubernetes cluster, he was consistently seeing more concurrent users than I was. It also makes sense that I was able to see more concurrent users with `arrivalRate` if that's the case, since you said it uses resources better.
 
> Latest is encouraged, so: 2.0.0-27 and 0.2.1 respectively.

I will update the versions in our repository to latest!

Thanks again! I look forward to any findings you gather regarding the issue I saw with artillery-engine-playwright and multiple concurrent virtual users.

**Reply by juanig1** [reply]
No problem!

Patch is [here](https://github.com/artilleryio/artillery/pull/1724), but still needs to be reviewed merged and a new version created, so it may take a couple of days for it to be live (assuming everything goes well).
There's a small summary in the PR but: the affected scenarios are when using `arrivalCount` or when `arrivalRate` is less than the workers/cores of your machine.

As a work around until then, these are the easier approaches I can think of right now: 
1) Running with a fixed amount of workers: `WORKERS=3 artillery run script.yaml` (3 because its the `arrivalRate` on your original script)
If not:
2) Running: `DEBUG=artillery:worker artillery run script.yaml`. This will print each worker config in the CLI (and you are going to see the bug in distributing it first hand 😄, as they wont add up to the total in you script). Then, bumping you `maxVusers` (and rerunning) until you get a total of 10 between all threads 
3) A mix of 1) and 2) if needed.

Let me know how it goes, hope to get that new version out soon.
Thanks again for the report!

**Reply by joaormouta** [reply]
Hi all.
When is it going to be available? :) I saw the patch 🥇 on github but not in x.27 version :(

**Reply by JulieMatic** [reply]
@juanig1 Can you explain the relationship between `WORKERS`, `arrivalRate`, and `maxVusers`? Using the workaround of setting `WORKERS=10`, we are able to run our tests with the Playwright portion executing in a stable manner, but we get a lot more users than we expected. I'm not sure the exact phase configuration, but at one point, my coworker was seeing over 20 concurrent users with `maxVusers: 10` (we want to cap at 10 users during the day so that we don't put too much strain on the environment when customers are mostly likely to be active). 

Based on his experiments, he said it looked like adding more workers would lower the effective arrival rate but made the test much more likely to exceed `maxVusers`.

**Reply by juanig1** [reply]
@joaormouta sorry, just seeing your message right now:
we are aiming for 2.0.0-28 to be released early next week if everything goes well. There are a couple of other enhancements that still need a bit of tweaking.

@JulieMatic Sure!
Internally, each phase workload is split between n `workers` (by default one per core, can be overridden) to better use resources. Ideally it _should not_ affect tests, but realistically _it does_ as resources are finite.
Also as work is distributed, there may be certain differences with different amount of workers (but we are actively working on getting that ironed out!).

`arrivalRate` is the amount of virtual users your phase generates per second.

`maxVusers` is the maximum amount of virtual users at any given time of the test.

Even though these last two sound similar, setting maxVusers may be useful if, for example, the service you are testing takes longer than 1s to respond. In those cases you may end up with more concurrent requests than the `arrivalRate` you set: as new vusers are created while older vusers are still being processed on your service.

If I did a good job explaining `workers`, `maxVusers`, and `arrivalRate`, you may now make some sense out of `DEBUG=artillery:worker artillery run script.yaml` output. It should show each worker configuration, if that helps.
If not, feel free to ask away!

Oh, still worth mentioning that we got this [fix](https://github.com/artilleryio/artillery/pull/1724) merged based on your report, but will only be available once 2.0.0-28 ships.
So bear in mind you may be getting a weird `maxVusers` distribution until that update is live (both arrivalRate and `maxVusers` should be distributed evenly across all `workers`).

**Reply by JulieMatic** [reply]
@juanig1 Thanks! So to clarify, when my coworker set `maxVusers` to 10, he should have never seen the concurrent users go above 10, regardless of what `WORKERS` is set to? Another coworker thought he had seen docs somewhere that implied the number of concurrent users would be something like `WORKERS` * `maxVusers`, but we couldn't find anything when we looked yesterday.

I will mention the `DEBUG` command to my coworker so he can try it out next time he runs our tests.

Do you have any recommendations for what a good phase configuration for our scenario would be once 2.0.0-28 ships? Once we upgrade, we would prefer to leave `WORKERS` at default, and our Playwright script probably takes about 60 seconds. Given that we want to set `maxVusers` to 10, what should we set `arrivalRate` to so that we don't see weird distributions across workers?

**Reply by juanig1** [reply]
> my coworker set maxVusers to 10, he should have never seen the concurrent users go above 10, regardless of what WORKERS is set to?

Exactly!
If we can revisit this with the newer version, that would be great; just to avoid debugging something that may be already fixed.
Still, if you have an scenario which shows weird behavior and its reproduce-able, feel free to share a simple version of it.

> concurrent users would be something like WORKERS * maxVusers

Hmm, should not be in any doc, but still this was the case up until recently.
We patched it to follow what is set in the script and to further  decouple test behavior from worker count. 
This is the part that's getting [fixed](https://github.com/artilleryio/artillery/pull/1724) in the next version

> Do you have any recommendations for what a good phase configuration for our scenario would be once 2.0.0-28 ships?

Workers on default should be fine!
`arrivalRate` and `maxVusers` depends on your desired scenario and what you are testing:
  - `arrivalRate` set to whatever amount of 'new flows per second' you want
  - `maxVusers` set to ten if that's what you are aiming to emulate/test
  - "weird distributions" of `maxVusers` should be fixed by then!
Just thing of the scenario you want to run and 'translate it to these terms (_"A maximum of `maxVusers` concurrent users trying to make `arrivalRate` requests per second for `duration` seconds"_)

**Reply by JulieMatic** [reply]
Hi @juanig1 . I was wondering if there is an ETA for version 2.0.0-28? On artillery.io, I see a note saying that there was an update on December 26, but if I look at Github, it still shows the latest version as 2.0.0-27 from November 16.

**Reply by juanig1** [reply]
Hey, happy holidays.
Sorry for the delay.

We haven't pushed the new version yet. Will be done once hassy is back from PTO. Should be out late this week or the next one

**Comment by JulieMatic** [comment]
@juanig1 Have you had a chance to investigate the issue with artillery-engine-playwright? 

Applying the workaround of setting `WORKERS=3`, I am now seeing the issue I detailed in my original post in section 'Only the "active" Chrome window is interacted with' when I'm using `artillery version 2.0.0-27` and `artillery-engine-playwright version 0.2.1`. 

I've seen it in both headless and headed mode on my laptop, and my coworker saw it running headless in a Kubernetes cluster (to my knowledge, he didn't try headed). Our working theory is that using the default 1 worker, the limited resources on my laptop prevented actual concurrency, which is why I didn't originally see the issue in the latest versions.

Here is the relevant output of `npm list` from my coworkers attempt yesterday:

> +-- @playwright/test@1.26.0
+-- artillery-async@1.0.5
+-- artillery-engine-playwright@0.2.1
+-- artillery-engine-sql@1.0.1
+-- artillery-engine-tcp@1.0.0
+-- artillery-plugin-expect@2.1.0
+-- artillery-plugin-publish-metrics@2.0.2
+-- artillery-plugin-statsd@2.2.1
+-- artillery@2.0.0-27
`-- typescript@4.8.3

I also tried locally with the same versions, except using `@playwright/test@1.25.2` but it didn't make a difference (only 1 or 2 of my Chrome windows filled out the login fields and logged in correctly, and the others got stuck without interacting with the window). 

Please let me know if I should spin this off into a new discussion, since the 2 issues I posted seem to be quite different in hindsight. Thanks!

### Update
We believe we found a workaround for the issue with Playwright interacting only with the "active" window:
1. Upgrade `@playwright/test` to version `1.28.1` 
2. Run artillery like `WORKERS=10 artillery run script.yaml`

With the above workaround and setting `MaxVusers: 10` in our YAML file, we seem to be getting 10 concurrent users and all 10 users interacting with their window as expected. 

Upgrading Playwright alone wasn't enough to fix the issue. I'm not sure if setting `WORKERS=10` without upgrading Playwright would have worked, but in our case, we're fine with upgrading Playwright to `1.28.1`

**Reply by juanig1** [reply]
Hi, glad you found a workaround, looking at this right now.

**Reply by juanig1** [reply]
Hey, can't reproduce this on my machine, but this may be on me. Can you share a simple version of your script/processor?

**Reply by JulieMatic** [reply]
Hi @juanig1, I think maybe the issue has changed a bit since my original post. 

Yesterday after I saw your message that you couldn't reproduce the issue, I tried to reproduce with a simple script just going to google.com and typing in the search box. I wasn't able to get this simple test to fail, but when I ran with the most users my laptop could handle (I used about 18 concurrent users because at this point moving my mouse was starting to get laggy), I noticed that some windows took a very long time to click and type into the search box. Because I had set the test timeout to 90 seconds, the simple test did eventually catch up and finish, but my actual tests probably take about 60 seconds when things go well, so they were reaching the 90 second timeout when my machine was under load.

At this point, I don't believe I'm seeing the issue where 2 virtual users are interacting with the same window, and the issue now is that when the machine is reaching its resource limits, the tests are very slow.

**Reply by juanig1** [reply]
Yeah, hitting the resource limit of the machine may lead to unexpected behavior like that.
Thank you for taking the time to test!
Edit: If you are using AWS, [running on lambda](https://www.artillery.io/blog/open-source-distributed-load-testing-with-lambda) may be a nice alternative!

**Reply by hassy** [reply]
Great to hear that the original bug is gone!

The slowdown with multiple Chrome instances is to be expected, as each instance consumes quite a bit of memory. Testing at production-scale would always require running those tests from the cloud.

By the way, the Lambda support in Artillery does not work for Playwright yet. It's possible but not straightforward to run Playwright inside a Lambda, and we haven't implemented support for it yet.

**Comment by mjohnson767** [comment]
Hello Hassy, I'm very interested in the Artillery | Playwright testing but experiencing the same issue as listed above. 
Build: artillery/2.0.0-37 win32-x64 node-v18.14.0
My goal is run tests with the VUser concurrent limit with a ramp-up (to the limit) and duration for the test. This is pretty much standard for all load testing tools. For some quick testing I've set a low amount of Vusers to test it out. With the following setup the actual concurrent rate is only 1 VUser that just repeats until the duration is hit. 

config:
  target: "https://mytesturl.com"
  phases:
    - duration: 60
      rampTo: 3
      name: Ramp up load  
      maxVusers: 3
  payload:
    - path: ./logins.csv
      fields:
        - "userid"
        - "recordid"
      order: sequence
 
If I take out the maxUsers I'll see several browsers open with only one of the actually going through the workflow. 

Attempting to run with the WORKERS setting fails with this version. WORKERS=3 artillery run script.yaml
WORKERS=3 : The term 'WORKERS=3' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, 
verify that the path is correct and try again.

Do you have any ideas? 
I've ran tests with Locust/Selenium locally and I know I can 20-30 browsers on my laptop. Running Artillery/Playwright should easily out perform locust/selenium. With a single load generator we were able to run 175 headless browsers. I'm hoping to surpass that with Artillery/Playwright.

**Comment by cgrant48** [comment]
Hi, I am also looking into using artillery | playwright testing and I believe I may be running into a similar obstacle. I am attempting to get at least 300 concurrent users running but I am continuously running into timeout errors. I have it set for each user to be able to login with username and password then press login. After about 30 users I begin to see the timeout errors populate and the waiting for page.locator to be visible error shows up as well.

Currently this is what my YAML looks like.
config:
  target: "https://testurl.com" 
    phases:
    - name: Warmup Login
      duration: 600
      arrivalCount: 300
       # Enable the Playwright engine:
  engines:
    playwright:
      launchOptions:
        headless: true
        args: ["--headed"]
  processor: "./javascripts/automation.js"
  
payload:
  path: "./users.csv"
  order: sequence
  loadAll: true
  skipHeader: true
  delimiter: ","
  skipEmptyLines: true
  fields:
    - "username"
    - "password"

If I am not mistaken that is asking for 300 users to be present by the end of 10 minutes. I apologize if this isn't enough information to go on. If You need further explanation I would be happy to provide it. I have added some timeout delays into my js file thinking that may help, but I am not sure if it is.

**Reply by up201306506** [reply]
I believe that the arrivalCount specifies the total amount of users, and not the total amount of "concurrent" users. If you wanted exactly 300 executions of the login, then what you have is correct. If you wanted 300 Vu's running at the same time, then I don't think you do.

You can quickly verify it by creating a simple test with the following environment:
```
    quicktest:
      phases:
        - name: hello
          arrivalCount: 3
          duration: 30
```

Which should yield both a vusers.completed and vusers.created of 3 on the final report, and a single one for each period.

To have 300 concurrent, you will have to set maxVusers to 300 and then have an arrival rate faster than the consumption. What the original post had was the ideal manner.

With that said, part of the reason why you're getting timeouts might be just resources. Running anything more than a mere couple playwright vUsers (and you-re doing it in the hundreds) on a local machine is generally a disastrous idea because all your resources will get hogged and then the vUsers themselves will slow down to a crawl, and then you'll hit time outs regardless of what's actually happening.

**Reply by mjohnson767** [reply]
Are there any benchmarks of running with the number of VUsers per/server? It would be good to know the limitations that you've discovered so that when running on a local server or using the cloud service it would be helpful for planning. I know the other limitations of the other tools running on load generators (servers with 32 Gigs ram running on a GIG network). 
Silk Performer (now discontinued): 12 users
Trueclient: 45-50 Users
Locust/selenium: 175 
XK6: unknown - decided not to go that way due to the many limitations of the tool
Artillery/Playwright: Mentioned that the tool supports thousands of users but how many users per/server. If I wanted to run a 300 user test for an hour would I need two servers? Benchmarks would be very helpful here.

**Reply by up201306506** [reply]
Sorry man, I don't have any specific benchmarks I could give beyond "My machine crashes with just 10 users on this script, but not with this one", I'm still relatively new looking into this. Part of why I was even on here was to find out the recommended way to measure the current amount of vusers because I want to measure this stuff too. 

I am surprised you're managing so many Selenium drivers concurrently. And I also abandoned k6 a couple weeks back when I saw how limited the browser API were.

**Reply by bernardobridge** [reply]
@up201306506 , thank you for your contributions in the Discussions, that was really well put!

---

@mjohnson767 / @cgrant48 , Playwright scripts can be quite complex/vary a lot and depend a lot on the nature of what your application and script are doing. We don't have benchmarks yet for that reason, and what we advise would be: 
- use our open-source distributed load testing solution ([Fargate](https://www.artillery.io/docs/load-testing-at-scale/aws-fargate)) to run these tests;
- do some trial/error by putting your individual test script with as low as VU as possible, so a single fargate container won't have issues (you could also do some of this testing locally, but bear in mind that the Fargate resources will likely be lower per machine);
- you can also use `--launch-config` to change [cpu and memory](https://www.artillery.io/docs/load-testing-at-scale/aws-fargate) limits in each Fargate container, which can help get more per container;
- then you can increase the `--count` to scale up your test to your needs. So for example, if you see a single fargate container can handle 5 VU, but you need 50 VU, you would set `--count 10`. _Note: you may need to request AWS to increase limits if your count is high enough, but that's a simple support ticket._

Hope that helps!

**Reply by cgrant48** [reply]
@bernardobridge I will definitely be looking into fargate. I do not know much about this open source solution but from the sound of it, this is probably just the thing I have been looking for. 

Thank you!

**Comment by mjohnson767** [comment]
Here's some results In doing my own benchmarking with a pretty simple script that runs through 4 actions of:
Launch page
Login
Search for a record
logout
Application was built using Azure and building the test with Any protocol level load testing tool is not possible. 
Tested with Settings of no pauses between each transaction and also a set of tests with 7 second pauses between each transaction. 
Tested on some well-equipped servers. 
Results: Running 10 VUsers per/server is the max peak performance.
Anymore VUsers and the CPU of the (load generator) server climbs and the response times climb as a result.
I was able to run up to 30 VUusers on a server but the transaction response times went from 3 seconds to 65 seconds. 
Running from a laptop with with 32 GIG memory, i7 6 core proc I was able to run 15 VUsers but I would run into OutofMemory errors. CPU was pegged at 100%.
Running the test(s) with pauses in the script made little difference in the CPU consumption. Additionally, Playwright ran through the scenario very quickly and with no issues. 
I know running every VUser in it's own thread was an important concept when building the integration with Playwright but the whole idea changes when you can only run a small handful of VUsers per/server and the response time climb because of the tool and not the load on the application. 
The Locust tool running selenium uses GEvents to run the browser processes and was able to run 175 Browser sessions per/server without any response degradation. Playwright is a solid tool and the rapid script development makes it the preferred choice for this purpose, especially with Artillery. But, to make it an effective approach you'd need a 5-10x improvement in running browsers per/server, with or without AWS and without any load generator CPU/memory issues.

**Reply by hassy** [reply]
Thanks @mjohnson767, appreciate you taking the time to run some benchmarks, that's very useful!

In general we expect users to scale their Playwright-based load tests with Fargate. Artillery makes it really easy to spin up hundreds of containers on Fargate without any extra work on the user's part. Running production-like load tests from a single machine - even if it's possible to generate enough load - is generally not a good idea.

That said, there are some optimizations we're considering, such as reusing browser contexts between VUs. At the moment every VU is an entirely separate instance of Chromium.

I don't have much experience with Selenium, but if it uses headless Chromium underneath, there shouldn't be much discernible difference in terms of CPU/memory overhead between it and Playwright. Making a guess here - higher concurrency on Selenium + Locust is probably due to reuse of a single browser instance between VUs.
